src/any_agent/callbacks/span_generation/llama_index.py

```diff
@@ -15,43 +16,95 @@
 
 class _LlamaIndexSpanGeneration(_SpanGeneration):
     def before_llm_call(self, context: Context, *args, **kwargs) -> Context:
-        llm_input: list[ChatMessage] = args[1]
-        input_messages = [
-            {
-                "role": message.role.value,
-                "content": message.content or "No content",
-            }
-            for message in llm_input
-        ]
-        model_id = context.shared["model_id"]
+        # Handle direct dict format (from call_model wrapper)
+        if "messages" in kwargs and isinstance(kwargs["messages"], list):
+            input_messages = kwargs["messages"]
+            if not input_messages:
+                return context
+            model_id = kwargs.get("model", "No model")
+        # Handle LlamaIndex callback format
+        elif len(args) >= 2 and isinstance(args[1], list):
+            llm_input: list[ChatMessage] = args[1]
+            input_messages = [
+                {
+                    "role": message.role.value,
+                    "content": message.content or "No content",
+                }
+                for message in llm_input
+            ]
+            model_id = context.shared["model_id"]
+        else:
+            return context
 
         return self._set_llm_input(context, model_id, input_messages)
 
     def after_llm_call(self, context: Context, *args, **kwargs) -> Context:
-        agent_output: AgentOutput = args[0]
-
-        output: str | list[dict[str, Any]] = ""
-        if response := agent_output.response:
-            if content := response.content:
+        response = args[0]
+        token_usage: Usage | None
+        # Handle litellm ModelResponse (from call_model wrapper)
+        if hasattr(response, "choices") and hasattr(response, "model_extra"):
```

Comment:
Doesn't need to be part of this (thanks for already refactoring some stuff) but this is probably a good candidate to split into separate methods like `_handle_litellm_response` and `_handle_llamaindex_response`